interfaces/builtin/polkit.go

```diff
@@ -0,0 +1,156 @@
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/polkit"
+	"github.com/snapcore/snapd/polkit/validate"
+	"github.com/snapcore/snapd/snap"
+)
+
+const polkitSummary = `allows access to polkitd to check authorisation`
+
+const polkitBaseDeclarationPlugs = `
+  polkit:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const polkitBaseDeclarationSlots = `
+  polkit:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const polkitConnectedPlugAppArmor = `
+# Description: Can talk to polkitd's CheckAuthorization API
+
+#include <abstractions/dbus-strict>
+
+dbus (send)
+    bus=system
+    path="/org/freedesktop/PolicyKit1/Authority"
+    interface="org.freedesktop.PolicyKit1.Authority"
+    member="{,Cancel}CheckAuthorization"
+    peer=(name="org.freedesktop.PolicyKit1", label=unconfined),
+dbus (send)
+    bus=system
+    path="/org/freedesktop/PolicyKit1/Authority"
+    interface="org.freedesktop.DBus.Properties"
+    peer=(name="org.freedesktop.PolicyKit1", label=unconfined),
+dbus (send)
+    bus=system
+    path="/org/freedesktop/PolicyKit1/Authority"
+    interface="org.freedesktop.DBus.Introspectable"
+    member="Introspect"
+    peer=(name="org.freedesktop.PolicyKit1", label=unconfined),
+`
+
+type polkitInterface struct {
+	commonInterface
+}
+
+func (iface *polkitInterface) getActionPrefix(attribs interfaces.Attrer) (string, error) {
+	var prefix string
+	if err := attribs.Attr("action-prefix", &prefix); err != nil {
+		return "", err
+	}
+	if err := interfaces.ValidateDBusBusName(prefix); err != nil {
+		return "", fmt.Errorf("plug has invalid action-prefix: %q", prefix)
+	}
+	return prefix, nil
+}
+
+func loadPolkitPolicy(filename, actionPrefix string) (polkit.Policy, error) {
+	content, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf(`could not read file %q: %v`, filename, err)
+	}
+
+	// Check that the file content is a valid polkit policy file
+	actionIDs, err := validate.ValidatePolicy(bytes.NewReader(content))
+	if err != nil {
+		return nil, fmt.Errorf(`could not validate policy file %q: %v`, filename, err)
+	}
+
+	// Check that the action IDs in the policy file match the action prefix
+	for _, id := range actionIDs {
+		if id != actionPrefix && !strings.HasPrefix(id, actionPrefix+".") {
+			return nil, fmt.Errorf(`policy file %q contains unexpected action ID %q`, filename, id)
+		}
+	}
+
+	return polkit.Policy(content), nil
+}
+
+func (iface *polkitInterface) PolkitConnectedPlug(spec *polkit.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	actionPrefix, err := iface.getActionPrefix(plug)
+	if err != nil {
+		return err
+	}
+
+	mountDir := plug.Snap().MountDir()
+	policyFiles, err := filepath.Glob(filepath.Join(mountDir, "meta", plug.Name()+".*.policy"))
```

Comment:
we use subdirectiories of meta for some things, wondering if this should be meta/polkit/*?  we have meta/gui as well already but not sure polkit is GUI only?